src/pip/_internal/operations/prepare.py

```diff
@@ -427,18 +430,26 @@ def _fetch_metadata_using_lazy_wheel(
             )
             return None
 
+        if link.netloc in self._domains_without_http_range:
+            return None
+
         wheel = Wheel(link.filename)
         name = canonicalize_name(wheel.name)
         logger.info(
             "Obtaining dependency information from %s %s",
             name,
             wheel.version,
         )
-        url = link.url.split("#", 1)[0]
         try:
-            return dist_from_wheel_url(name, url, self._session)
-        except HTTPRangeRequestUnsupported:
-            logger.debug("%s does not support range requests", url)
+            return dist_from_wheel_url(name, link.url_without_fragment, self._session)
+        except HTTPRangeRequestUnsupported as e:
+            logger.debug(
+                "domain '%s' does not support range requests (%s): "
+                "not attempting lazy wheel further",
+                link.netloc,
+                str(e),
+            )
+            self._domains_without_http_range.add(link.netloc)
```

Comment:
I see how you make this change for `HTTPRepository` for python-poetry/poetry#8815; for pip, I am looking to further cache metadata lookups in #12256, which I believe will obviate the need for this check (since once the metadata is downloaded once, it will . In fact, please make sure also to look at #12257 and #12258 if you haven't already, as I believe poetry may be able to take advantage of at least the HTTP caching directives from #12257, if not #12258 (I don't know if you resolve against concrete interpreters the same way)?